public/app/features/dashboard/components/GenAI/hooks.ts

```diff
@@ -0,0 +1,95 @@
+import { useState } from 'react';
+import { useAsync } from 'react-use';
+import { Subscription, finalize } from 'rxjs';
+
+import { llms } from '@grafana/experimental';
+
+// TODO: Add tests
+export function useOpenaiStreamHook(
+  model = 'gpt-3.5-turbo',
+  temperature = 1
+): {
+  setMessages: React.Dispatch<React.SetStateAction<llms.openai.Message[]>>;
+  reply: string;
+  started: boolean;
+  finished: boolean;
+  loading: boolean;
+  error: Error | undefined;
+  value:
+    | {
+        enabled: any;
+        stream?: undefined;
+      }
+    | {
+        enabled: any;
+        stream: Subscription;
+      }
+    | undefined;
```

Comment:
I'm not super sure about this one 🤔  why don't we just return two more properties: `enabled` and `stream`?

Reply:
+1 to @ivanortegaalba comment.

Reply:
I tried to remove this and it's still done because, basically, react's useAsync hook requires it. It returns 'loading' (which refers to the useAsync itself, not to the contained logic, which is very confusing in my opinion -- that we have more than one ), error (which I believe will only be defined if the function in useAsync has thrown an error), and "value", which can be an object. So, all of our custom values end up within "value".

Also: Returning the stream is, to my eye, totally unnecessary, since we do not do anything with it after returning it. It is left as-is because Typescript flags it as unused otherwise, I believe inaccurately. It is possible that returning it helps with a scope issue such that the stream can't populate 'reply' unless it's returned from the useAsync hook, but I would find that counterintuitive.

In a big picture way: useAsync is what we're using as a bridge between rxjs (which we are using for stream management) and react state, and it's opinionated about this.